### PR TITLE
Changes how the ids and positions are retrieved

### PR DESCRIPTION
**Pull request by sergiorg-hpc**
The PR alters how the node IDs and positions are retrieved from storage. The node pointers are now filtered into a sub-map and used directly for different purposes, instead of splitting the IDs and positions like in the original code. The IO pattern for the IDs is also different and avoids the use of multiple hyperslabs, following a similar approach to what was introduced when fetching the actual data from storage (https://github.com/BlueBrain/libsonata/pull/104). In addition, the changes simplify and unify parts of the code (e.g., removing the use of lambda-functions).

Using a reference report of around 60TB, fetching one single timestamp from the file for all the GIDs would imply the following:
- In the former implementation, getting the metadata of each of the 4,234,929 GIDs of the report implies triggering a hyperslab + read for around 300 values per compartment (i.e., 4,234,929 reads of around 300 values each).
- In the proposed implementation, only a single hyperslab is triggered to fetch the data for the 4,234,929 IDs. Then, this data is conveniently filtered in memory.

These millions of small reads from the original implementation were causing problems recently in the GPFS file system of BB5, specially when scientists were launching multiple processes over multiple nodes conducting similar requests on big reports.

### CODE REVIEW

**Comment by mgeplf** [comment]
> Using a reference report of around 60TB with a single process, fetching one single timestamp from the file for all the GIDs takes the following:

How does the change impact fetching multiple timesteps w/ a subset of GIDS?

**Comment by sergiorg-hpc** [comment]
> How does the change impact fetching multiple timesteps w/ a subset of GIDS?

@mgeplf If the GIDs are quite far from each other (e.g., 2 GIDs, one in the very beginning and one at the very end), then the performance can be worse if the report is extremely large. However, the data will always be fetched following the same technique, meaning that even if we leave the implementation as it is, the cost to fetch the data will still be worse in this particular edge case. This was the trade-off we made last year when we switched from having the millions of hyperslabs to just one hyperslab per timestep. Overall, it provides lots of performance benefits in comparison.

But just to clarify, my next work item would be in that direction actually! My plan is to have a threshold that would create ranges of data selection, instead of a single range like now. This way, following the previous example, if the gap between the GIDs is large, then we would just create two (or more) separate hyperslabs with as much data as possible. This would still be optimal to the file system because we have fewer requests, while still handling the edge cases that can happen.

However, this is a separate task. For now, I wanted to unify both implementations, merge this into `libsonata` and then the next step would be to add this extra step. Both Joseph and Andras have tried this patch and have obtained clear performance benefits.

What do you think?

**Comment by mgeplf** [comment]
> But just to clarify, my next work item would be in that direction actually!

Cool.

Can we at least run the `fetching multiple timesteps w/ a subset of GIDS` w/ the old code, and the new code, and put that in this ticket?

**Comment by sergiorg-hpc** [comment]
Yup, I will run a test throughout the morning and add it here, no problem at all.

Just an additional note that I forgot to mention. Keeping the old implementation is heavy for GPFS when multiple processes open reports. For instance, Joseph was involuntarily blocking GPFS for everybody with his simulations and most of it was just in the beginning fetching the IDs and positions. Regardless of whether we merge this or not, the data will still be fetched with the `min`/`max` range for the data anyway. So, having the IDs fetched similarly is equivalent to adding an extra timestep to the report, if that makes sense (i.e., like 10001 `tsteps` instead of 10000).

Nonetheless, I 100% agree, let me just do a couple of runs just to understand where we are and we can take it from there.

**Comment by sergiorg-hpc** [comment]
After rebasing from `master` to fix the compilation mode bug (see #175), here are some results **using a single process** to fetch one timestep with the current version from `master` (i.e., `v0.1.11`) versus the same version with the changes from this PR (i.e., `v0.1.11+Patch`):
image

The figure illustrates Execution Time in seconds on the y-axis and the number of GIDs requested in the x-axys. The test script uses a reference ~60TB report and randomly selects GIDs from the input, mimicking some short of spatial bining. The tests ask for a single timestep, which would first trigger internally the fetching of the GIDs and the positions (i.e., the part that involves the changes from this PR) plus accessing the data to retrieve a single timestep (i.e., this part identical in both versions and previously optimized in #104*). The table at the bottom shows specifically the execution time for each test.

I must admit that, after the compilation fix, the original code doesn't look as bad as it used to. For instance, selecting chunks of sequential GIDs is not that worse. However, I would like to kindly point out that we cannot keep the code as it is, otherwise it will be quite stressful for GPFS when scientists utilize hundreds of processes on big reports. Moreover, this PR is required for the next optimization step that reads in chunks of a certain threshold, as I mentioned earlier in one of my comments.

For reference purposes, here is the Python script used for the tests:

```python
import libsonata
import time
import sys
import numpy as np

elements = libsonata.ElementReportReader('/gpfs/.../002/AllCurrents.h5')
population_elements = elements['All']
print("File loaded!")

for i in range(0, 7, 1):
    count = pow(10, i)
    node_ids = population_elements.get_node_ids()

    np.random.seed(21)
    node_ids = list(np.random.choice(node_ids, count, replace=False))

    start = time.time()
    data_frame = population_elements.get(node_ids, tstart=5000.0, tstop=5000.1)
    end = time.time()

    print(str(count) + " / " + str(end-start) + "\n-----")
```

\* Note that fetching more than one timestep in the tests will not illustrate any significant difference, as this part of the code is identical.

File: include/bbp/sonata/report_reader.h [hunk 1]
```diff
@@ -166,17 +167,17 @@ class SONATA_API ReportReader
         Population(const H5::File& file, const std::string& populationName);
         std::pair<size_t, size_t> getIndex(const nonstd::optional<double>& tstart,
                                            const nonstd::optional<double>& tstop) const;
+        std::tuple<NodePointers, Range, typename DataFrame<KeyType>::DataType> getElementIds(
```
Comment (matz-e) [inline include/bbp/sonata/report_reader.h:170]: I'd like some documentation here at some point.

**Review by matz-e — DISMISSED** [review]
Looks good to me, some top-level abstract documentation would be good, though.

File: include/bbp/sonata/report_reader.h [hunk 2]
```diff
@@ -166,17 +166,19 @@ class SONATA_API ReportReader
         Population(const H5::File& file, const std::string& populationName);
         std::pair<size_t, size_t> getIndex(const nonstd::optional<double>& tstart,
                                            const nonstd::optional<double>& tstop) const;
+        std::pair<NodePointers, Range> getNodePointers(
+            const nonstd::optional<Selection>& node_ids = nonstd::nullopt) const;
+        typename DataFrame<KeyType>::DataType getElementIds(const NodePointers& node_pointers,
+                                                            const Range& range) const;
```
[the remark this answers is not in the document]
Comment (matz-e) [reply include/bbp/sonata/report_reader.h]: Oops, sorry. This was still somewhere in the "cache"

File: include/bbp/sonata/report_reader.h [hunk 3]
```diff
@@ -98,13 +99,14 @@ class SONATA_API SpikeReader
     mutable std::map<std::string, Population> populations_;
 };
 
+using Range = std::pair<uint64_t, uint64_t>;
```
[the remark this answers is not in the document]
Comment (sergiorg-hpc) [reply include/bbp/sonata/report_reader.h:102]: Yes, sorry, this was actually a temporary change that was left here due to a compilation issue. It is fixed now by referencing the correct type.

File: include/bbp/sonata/report_reader.h [hunk 4]
```diff
@@ -166,17 +166,25 @@ class SONATA_API ReportReader
...
+         * for performance to understand how and where to retrieve the data from storage.
+         *
+         * \param node_ids limit the report to the given selection. If nullptr, all nodes in the
+         * report are used
+         */
+        std::tuple<NodePointers, Range, typename DataFrame<KeyType>::DataType> getElementIds(
+            const nonstd::optional<Selection>& node_ids = nonstd::nullopt) const;
 
-        std::map<NodeID, Range> nodes_pointers_;
+        NodePointers nodes_pointers_;
```
[the remark this answers is not in the document]
Comment (sergiorg-hpc) [reply include/bbp/sonata/report_reader.h:180]: As big as the number of GIDs times the ranges in the file (i.e., 3 x `uint64_t` per GID). This data structure is required and has not changed compared to the original implementation.

Keep in mind that, even when using a very big report like the reference 60TB, it would be around 100MB of memory in total (4234929 x 3 x 8 = 96.9MB). I'll reply to your other comment below, but the idea of this PR is to re-use the same data structure instead of splitting the `NodeID`s in a vector and then the positions in another vector, like in the previous implementation.

File: src/report_reader.cpp [hunk 5]
```diff
@@ -284,24 +285,64 @@ std::vector<NodeID> ReportReader<T>::Population::getNodeIds() const {
...
-                       nodes_pointers_.end(),
-                       std::back_inserter(node_ids),
-                       [](const std::pair<NodeID, Range>& node_pointer) {
-                           return node_pointer.first;
-                       });
-    } else if (selection->empty()) {
-        return {};
-    } else {
-        node_ids = selection->flatten();
+std::tuple<NodePointers, Range, typename DataFrame<T>::DataType>
```
[the remark this answers is not in the document]
Comment (sergiorg-hpc) [reply src/report_reader.cpp:288]: Agree!

File: src/report_reader.cpp [hunk 6]
```diff
@@ -284,24 +285,64 @@ std::vector<NodeID> ReportReader<T>::Population::getNodeIds() const {
...
+    // Helper function to update the min / max values and count
+    const auto& update_range = [&range, &eids_count](const Range& range_new) {
+        range.first = std::min(range.first, range_new.first);
+        range.second = std::max(range.second, range_new.second);
+        eids_count += (range_new.second - range_new.first);
+    };
+
+    // Take all nodes if no selection is provided
+    if (!selection) {
+        node_pointers = nodes_pointers_;
```
[the remark this answers is not in the document]
Comment (sergiorg-hpc) [reply src/report_reader.cpp:304]: Yes, yes it is. The original implementation splits the `NodeID` keys of the map into an `std::vector`, and then the positions into another vector. The current change unifies the data structures for simplicity.

We thought about using an `std::shared_ptr` of the `node_pointers_` map, allowing us to just point to the original when this is "all the GIDs" (constant time) or to a new map when it is not. However, the performance of the copy for the `std::map` here is almost negligible compared to the rest of the code and it would make the source code more difficult to read.

File: src/report_reader.cpp [hunk 7]
```diff
@@ -284,24 +285,64 @@ std::vector<NodeID> ReportReader<T>::Population::getNodeIds() const {
...
+        const auto& max = range.second;
+
+        std::vector<ElementID> element_ids;
+        auto dataset_elem_ids = pop_group_.getGroup("mapping").getDataSet("element_ids");
+        dataset_elem_ids.select({min}, {max - min}).read(element_ids);
+
+        ids.reserve(eids_count);
+
+        for (const auto& node_pointer : node_pointers) {
+            const auto& node_id = node_pointer.first;
```
[the remark this answers is not in the document]
Comment (sergiorg-hpc) [reply src/report_reader.cpp:333]: Good catch, thanks!

**Comment by sergiorg-hpc** [comment]
Thank you very much for the comments and suggestions, @mgeplf. @jorblancoa and I have addressed most of the hints and also verified that the performance still remains equivalent.

If you like this version, let's go ahead and merge it to create a new `libsonata` release for deployment.

Have a nice weekend and thank you once again!

File: src/report_reader.cpp [hunk 8]
```diff
@@ -284,24 +284,63 @@ std::vector<NodeID> ReportReader<T>::Population::getNodeIds() const {
...
+        result.range.first = std::min(result.range.first, range_new.first);
+        result.range.second = std::max(result.range.second, range_new.second);
+        eids_count += (range_new.second - range_new.first);
+    };
+
+    // Take all nodes if no selection is provided
+    if (!selection) {
+        result.node_pointers = node_pointers_;
+
+        for (const auto& node_pointer : result.node_pointers) {
```
Comment (mgeplf) [inline src/report_reader.cpp:304]: does `node_pointer` need to be a ref? it's POD, no?

File: src/report_reader.cpp [hunk 9]
```diff
@@ -423,7 +438,8 @@ DataFrame<T> ReportReader<T>::Population::get(const nonstd::optional<Selection>&
         off_t offset = 0;
         off_t data_offset = (timer_index - index_start) / stride;
         auto data_ptr = &data_frame.data[data_offset * n_ids];
-        for (const auto& position : positions) {
+        for (const auto& node_pointer : node_pointers) {
```
Comment (mgeplf) [inline src/report_reader.cpp:441]: when the full selection is used, we know that, and can avoid the full copy of the `node_pointers` map, which can be quite large, no?

File: include/bbp/sonata/report_reader.h [hunk 10]
```diff
@@ -163,20 +158,34 @@ class SONATA_API ReportReader
                                const nonstd::optional<size_t>& tstride = nonstd::nullopt) const;
 
       private:
+        struct ElementIdsData {
```
[the remark this answers is not in the document]
Comment (sergiorg-hpc) [reply include/bbp/sonata/report_reader.h:161]: Any suggestions?

File: include/bbp/sonata/report_reader.h [hunk 11]
```diff
@@ -163,20 +158,34 @@ class SONATA_API ReportReader
...
+            typename DataFrame<KeyType>::DataType ids;
+            std::map<NodeID, Selection::Range> node_pointers;
```
[the remark this answers is not in the document]
Comment (sergiorg-hpc) [reply include/bbp/sonata/report_reader.h:163]: It was called `node_pointers` and I kept it, just to clarify.

File: src/report_reader.cpp [hunk 12]
```diff
@@ -284,24 +284,63 @@ std::vector<NodeID> ReportReader<T>::Population::getNodeIds() const {
...
+            if (it != node_pointers_.end()) {
+                result.node_pointers.emplace(*it);
+                update_range(it->second);
+            }
+        }
+    }
+
+    // Extract the ElementIDs from the GIDs
+    if (!result.node_pointers.empty()) {
+        const auto min = result.range.first;
```
[the remark this answers is not in the document]
Comment (mgeplf) [reply src/report_reader.cpp:321]: Ah, I must have missed them, where are they?

File: src/report_reader.cpp [hunk 13]
```diff
@@ -284,24 +284,63 @@ std::vector<NodeID> ReportReader<T>::Population::getNodeIds() const {
...
-                       nodes_pointers_.end(),
-                       std::back_inserter(node_ids),
-                       [](const std::pair<NodeID, Range>& node_pointer) {
-                           return node_pointer.first;
-                       });
-    } else if (selection->empty()) {
-        return {};
-    } else {
-        node_ids = selection->flatten();
+    ElementIdsData result;
```
[the remark this answers is not in the document]
Comment (sergiorg-hpc) [reply src/report_reader.cpp:289]: It is correct to return an empty selection. See line #416.
Again, the code was exactly like this before.

File: src/report_reader.cpp (hunk 9 above)
Comment (sergiorg-hpc) [reply src/report_reader.cpp:441]: As I explained in my previous comments, this PR does not change the behaviour of the code. The code used to copy all the node IDs into a vector, and then all the positions into a separate vector. The memory consumption increases in the current code.

The PR simplifies this by just using the same data structure instead.

Comment (mgeplf) [reply src/report_reader.cpp:441]: Can't we make the improvement here though?

File: src/report_reader.cpp (hunk 8 above)
Comment (mgeplf) [reply src/report_reader.cpp:304]: sorry, misread.

**Comment by sergiorg-hpc** [comment]
@mgeplf With all due respect, I think that we are missing the point here. Once again: the current `libsonata` reads the data of the reports with the `min`/`max` optimization. This is no different and will also be penalized when fetching the first and last GID in the old implementation.

This current PR unifies the same IO pattern for the metadata that is required to obtain the data. Without this change, the code triggers millions (I insist, millions) of small IO requests on big reports, just before fetching a single timestep of the data. The performance numbers listed above is to fetch a single timestep, just one! Then, you have to add the time to fetch the other 9999 timesteps of the report, that will have identical cost in both implementations.

I will upload now some of your suggested changes, ok?

**Comment by mgeplf** [comment]
> This PR unifies the same IO pattern for the metadata that is required to obtain the data.

I think this actually pessimizing some use cases, though; for instance if I use `get_node_id_element_id_mapping(sel)` on a selection of the large report, the time to get the result goes from about 6s to almost 10s.

**Comment by sergiorg-hpc** [comment]
> > This PR unifies the same IO pattern for the metadata that is required to obtain the data.
> 
> I think this actually pessimizing some use cases, though; for instance if I use `get_node_id_element_id_mapping(sel)` on a selection of the large report, the time to get the result goes from about 6s to almost 10s.

Can you provide the scripts and report you used? Also, I repeatedly mentioned that this PR is needed in order to include an optimization for the edge cases, while handling the big reports.

We have already optimized the code in 4-5 contributions, and this one adds to the previous contributions.

**Comment by mgeplf** [comment]
The test case is:
```
import libsonata
p = '/gpfs/bbp.cscs.ch/data/scratch/proj85/lotsofSims/baseline/633cbccb-f6b2-4e10-8f83-0a2655dcbcc7/000/AllCurrents.h5'
r = libsonata.ElementReportReader(p)['All']
mapping = r.get_node_id_element_id_mapping(libsonata.Selection([4228232, 15872]))
print(len(mapping))
print(mapping[0], mapping[-1])
```
The reason this is slower is likely due to `getNodeIdElementIdMapping` calling `getElementIds` which is doing more work than before.

The behavior also has changed (which we have to be careful about b/c of Hyrum's law): since there is a map now, the returned values are different:
Before, the second print gave: `[4228232, 0] [15872, 176]`, and now it gives: [15872, 0] [4228232, 579]`; so we should mention that the returned value isn't guaranteed to be stable.

**Comment by matz-e** [comment]
Seems to me that the test case is fetching two well-separated IDs, which is a known regression to be followed up and optimized on. It just seemed that doing it all in one go would lead to a more complex PR and breaking the work up would be better.

With the return values… that could be a costly re-sorting according to the user input. Seems doable, though.

File: src/report_reader.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (matz-e) [reply src/report_reader.cpp:321]: Benchmark scripts at the bottom of the post here: https://github.com/BlueBrain/libsonata/pull/174#issuecomment-1018826442

[the remark this answers is not in the document]
Comment (mgeplf) [reply src/report_reader.cpp:321]: Thanks - I just thought there were others where the other functionality that was impacted was also checked.

File: src/report_reader.cpp (hunk 9 above)
Comment (mgeplf) [reply src/report_reader.cpp]: I'm going to give it a go; I will report back what I get.

File: include/bbp/sonata/report_reader.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (mgeplf) [reply include/bbp/sonata/report_reader.h:161]: How about `NodeIdElementLayout`; I will make the change.

**Comment by mgeplf** [comment]
> Seems to me that the test case is fetching two well-separated IDs, which is a known regression to be followed up and optimized on.

Yeah, fair enough, we just need to be clear on what is improving, and this seems like a reasonable tradeoff.

> It just seemed that doing it all in one go would lead to a more complex PR and breaking the work up would be better.

Yup, for sure.

File: src/report_reader.cpp (hunk 9 above)
Comment (mgeplf) [reply src/report_reader.cpp]: I tried switching to a map here:
https://github.com/BlueBrain/libsonata/compare/sandbox/srivas/ids_perf_improvements...sandbox/gevaert/ids-perf-improvements?expand=1

The case of fetching the full report is faster, so I think this should be pursued:
```
Map:
#IDS    | 1         | 2         | 3
      1 | 0.021234  | 0.004357  | 0.004069
     10 | 5.125821  | 4.981242  | 4.990284
    100 | 5.579101  | 5.563343  | 5.612758
   1000 | 5.994568  | 6.000875  | 6.436316
  10000 | 6.057319  | 6.065421  | 6.588673
 100000 | 6.722486  | 6.660799  | 7.208886
1000000 | 13.149290 | 13.744862 | 14.119598
4234929 | 37.245648 | 39.671034 | 39.629977

Vector:
#IDS    | 1          | 2         | 3
      1 | 0.005059   | 0.003865  | 0.003995
     10 | 5.104329   | 4.995507  | 5.019744
    100 | 5.963634   | 5.681968  | 5.617718
   1000 | 6.389871   | 6.375495  | 6.052416
  10000 | 6.196432   | 6.461813  | 6.104821
 100000 | 6.767185   | 7.148655  | 6.833382
1000000 | 12.229618  | 12.654732 | 12.594303
4234929 | 30.915542  | 32.286852 | 31.678779
```
Each column is a run, where a run loads the # of random IDS on the left; it does this for 3 full time frames, and averages the read time of a frame.

Comment (sergiorg-hpc) [reply src/report_reader.cpp]: Thank you very much, @mgeplf. Indeed, the `std::map` is no longer necessary and switching to a vector will make the memory accesses considerably faster.

I just squashed your branch, cleaned it up a bit, and pushed the changes into this branch. I will now verify that no major changes are observed in performance compared to the old version, just for the shake of completeness.

Comment (sergiorg-hpc) [reply src/report_reader.cpp]: Here is the commit:
https://github.com/BlueBrain/libsonata/pull/174/commits/e1dce9d50264aacd97766d0b875708567aeeee58

Comment (sergiorg-hpc) [reply src/report_reader.cpp]: The same evaluation of [fetching one GID of the figure](https://github.com/BlueBrain/libsonata/pull/174#issuecomment-1018826442) but with the update to `std::vector`, that also illustrates the same improvements that @mgeplf mentioned for large amounts of GIDs:

```
1 / 0.010879278182983398
10 / 5.343438386917114
100 / 5.943392753601074
1000 / 6.406604766845703
10000 / 6.451685428619385
100000 / 7.101217269897461
1000000 / 12.762017250061035
4234929 / 27.868659019470215
```

File: src/report_reader.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (sergiorg-hpc) [reply src/report_reader.cpp]: This has changed now, closing the conversation.

File: src/report_reader.cpp [hunk 14]
```diff
@@ -284,24 +284,63 @@ std::vector<NodeID> ReportReader<T>::Population::getNodeIds() const {
...
+        for (const auto& node_pointer : result.node_pointers) {
+            const auto node_id = node_pointer.first;
+            const auto& range = node_pointer.second;
+
+            for (auto i = (range.first - min); i < (range.second - min); i++) {
+                result.ids.emplace_back(make_key<T>(node_id, element_ids[i]));
+            }
+        }
+
+        // Reduces overhead of next IOps by clearing-up internal buffers of HDF5
```
[the remark this answers is not in the document]
Comment (sergiorg-hpc) [reply src/report_reader.cpp:354]: @jorblancoa and I replicated the issue on a separate example using the C library of HDF5. It seems to happen when you ask for a large hyperslab in a dataset and then move to another dataset in the same file where you also ask for another large range. For some reason, the first cold access takes a few seconds extra and we observed that fooling HDF5 does the trick. The old implementation did not show the issue because it was conducting small ~300-value accesses.

The plan to report this issue to the HDFGroup, [just like we did when we reported the ASCII bug that we thought it was our fault in `libsonata`](https://github.com/HDFGroup/hdf5/issues/544), if you remember. For now, let's keep it, please.

[the remark this answers is not in the document]
Comment (mgeplf) [reply src/report_reader.cpp:354]: > For now, let's keep it, please.

Yeah, for sure!

It's such an odd thing, but I also tried it w/ the 'old' implementation, and saw a bit of speedup (not as big as this PR, obv, but still non-zero).

Can you add the explanation you just gave as a comment (`It seems to happen...`)?  It would also be helpful to add you found it, so we can test it again in the future, and potentially drop it if it's fixed by upstream HDF5.

[the remark this answers is not in the document]
Comment (sergiorg-hpc) [reply src/report_reader.cpp]: Yes, I agree. I just updated the code with further clarifications.

File: src/report_reader.cpp [hunk 15]
```diff
@@ -284,24 +284,82 @@ std::vector<NodeID> ReportReader<T>::Population::getNodeIds() const {
...
+            for (auto i = (range.first - min); i < (range.second - min); i++) {
+                result.ids.emplace_back(make_key<T>(node_id, element_ids[i]));
+            }
+        }
+
+        // Temp. fix: When you ask for a large hyperslab in a dataset and then move
+        //            to another dataset in the same file where you also ask for
+        //            another large range, the next IOps take an extra few seconds.
+        //            We observed that fooling HDF5 hides the issue, but we should
+        //            verify this behaviour once new releases of HDF5 are available.
```
Comment (matz-e) [inline src/report_reader.cpp:358]: Nice!

**Review by matz-e — APPROVED** [review]

File: src/report_reader.cpp (hunk 9 above)
Comment (mgeplf) [reply src/report_reader.cpp]: cool; thanks.

**Comment by mgeplf** [comment]
Looks good to me; I asked @NadirRoGue about what the usage patterns were on the viz side, and it sounds like the usual number of node id's examined is `it can go from 100 to 200k`; so I think it's fine to pessimize the low end number for the gains in the larger views.

I can merge, or you can do the honours, @sergiorg-hpc.  Thanks for contribution!

**Review by mgeplf — APPROVED** [review]

**Comment by sergiorg-hpc** [comment]
> Looks good to me; I asked @NadirRoGue about what the usage patterns were on the viz side, and it sounds like the usual number of node id's examined is `it can go from 100 to 200k`; so I think it's fine to pessimize the low end number for the gains in the larger views.
> 
> I can merge, or you can do the honours, @sergiorg-hpc. Thanks for contribution!

Do not thank me, this was a team effort (props to @jorblancoa, @matz-e and others). Also, thanks to you and your team, @mgeplf.

I cannot merge because I do not have permissions, so please, you can go ahead. Also, if you don't mind, create a new tag `v0.1.11` so that I can create a PR in Spack to release the `Debug->Release` patch and this one on BB5 as soon as possible, ok? I will add you to the reviewers of that PR as well.

Once again, thank you for the insight and let's hope this fix can further help our users.
